File: be-nodejs/src/services/auth.service.ts
import { verifyHmacSha256 } from "@/libs/hashing";
import { mapUserToUserProfileResponse } from "@/libs/mappers/user.mapper";
import { User } from "@/models/user.model";
import { UserRepository } from "@/repositories/user.repo";
import {
      AuthResponse,
      LoginCredentials,
      RegisterData,
} from "@/types/auth.type";
import { UserProfileResponse } from "@/types/res/user.res";
import { JwtUtil } from "@/utils/jwt.util";

export class AuthService {

      private readonly userRepository: UserRepository;


      constructor() {
            this.userRepository = new UserRepository();
      }


      public async getProfile(accessToken: string): Promise<UserProfileResponse> {
            const decoded = await JwtUtil.verify(accessToken);
            const userId = decoded.id;
            const user = await this.userRepository.findById(userId);
            if (!user || !user.isEnable) {
                  throw new Error('User not found or inactive');
            }
            const userProfile: UserProfileResponse = await mapUserToUserProfileResponse(user);
            return userProfile;
      }


      public async refreshToken(refreshToken: string): Promise<{ accessToken: string }> {
            const decoded = await JwtUtil.verify(refreshToken);

            const user = await this.userRepository.findById(decoded.id);
            if (!user || !user.isEnable) {
                  throw new Error('User not found or inactive');
            }

            const tokenPayload = {
                  id: user.id.toString(),
                  email: user.email,
                  role: user.role,
            };
            const accessToken = await JwtUtil.generateAccessToken(tokenPayload);

            return { accessToken: accessToken };
      }


      public async authenticate(credentials: LoginCredentials): Promise<AuthResponse> {
            const user = await this.userRepository.findByEmail(credentials.email);

            if (!user || !user.isEnable) {
                  throw new Error("Invalid email or password");
            }
            const isPasswordValid = await verifyHmacSha256(credentials.password, user.password);

            if (!isPasswordValid) {
                  throw new Error("Invalid email or password");
            }
            const tokenPayload = {
                  id: user.id.toString(),
                  email: user.email,
                  role: user.role,
            };

            const accessToken = await JwtUtil.generateAccessToken(tokenPayload);
            const refreshToken = await JwtUtil.generateRefreshToken(tokenPayload);
            const authResponse: AuthResponse = {
                  userProfile: await mapUserToUserProfileResponse(user),
                  accessToken,
                  refreshToken,
            };

            return authResponse;
      }


      public async createAccount(registerData: RegisterData): Promise<AuthResponse> {
            console.log(registerData);
            const existingUser = await this.userRepository.findByEmail(registerData.email);
            if (existingUser) {
                  throw new Error("Email already exists");
            }

            const user = await this.userRepository.create({
                  email: registerData.email,
                  fullname: registerData.fullname,
                  password: registerData.password,
                  phone: registerData.phone,
                  dateOfBirth: registerData.dateOfBirth instanceof Date ? registerData.dateOfBirth : new Date(registerData.dateOfBirth),
                  role: registerData.role
            } as User);

            if (user != null) {
                  const tokenPayload = {
                        id: user.id.toString(),
                        email: user.email,
                        role: user.role,
                  };
      
                  const accessToken = await JwtUtil.generateAccessToken(tokenPayload);
                  const refreshToken = await JwtUtil.generateRefreshToken({
                        id: user.id.toString(),
                        email: user.email,
                        role: user.role
                  });
                  const authResponse: AuthResponse = {
                        userProfile: await mapUserToUserProfileResponse(user!),
                        accessToken,
                        refreshToken,
                  };
      
                  return authResponse;
            }

            throw new Error("An error occurred while creating the account");
      }

      public async getUserByToken(accessToken: string): Promise<User | null> {
            const decoded = await JwtUtil.verify(accessToken);
            return this.userRepository.findById(decoded.id);
      }

      public async getAllUsers(): Promise<UserProfileResponse[]> {
            const users = await this.userRepository.findAll();
            return Promise.all(users.map(user => mapUserToUserProfileResponse(user)));
      }

      public async updateUser(userId: string, updateData: Partial<User>): Promise<UserProfileResponse | null> {
            const updatedUser = await this.userRepository.update(userId, updateData);
            if (!updatedUser) {
                  return null;
            }
            return mapUserToUserProfileResponse(updatedUser);
      }

      public async getUserByEmail(email: string): Promise<UserProfileResponse | null> {
            const user = await this.userRepository.findByEmail(email);
            if (!user) {
                  return null;
            }
            return mapUserToUserProfileResponse(user);
      }

      public async updateUserByEmail(email: string, updateData: Partial<User>): Promise<UserProfileResponse | null> {
            const user = await this.userRepository.findByEmail(email);
            if (!user) {
                  return null;
            }
            const updatedUser = await this.userRepository.update(user.id, updateData);
            if (!updatedUser) {
                  return null;
            }
            return mapUserToUserProfileResponse(updatedUser);
      }

      public async deleteUserByEmail(email: string): Promise<boolean> {
            const user = await this.userRepository.findByEmail(email);
            if (!user) {
                  return false;
            }
            return await this.userRepository.delete(user.id);
      }

      public async updateUserStatusByEmail(email: string, isEnable: boolean): Promise<UserProfileResponse | null> {
            const user = await this.userRepository.findByEmail(email);
            if (!user) {
                  return null;
            }
            const updatedUser = await this.userRepository.updateStatus(user.id, isEnable);
            if (!updatedUser) {
                  return null;
            }
            return mapUserToUserProfileResponse(updatedUser);
      }
}

File: servers/IdentityService/Services/Auth/AuthService.cs
using IdentityService.Models;
using IdentityService.Web.Requests;
using IdentityService.Web.Responses;
using IdentityService.Repositories.User;
using IdentityService.Utils;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;

namespace IdentityService.Services.Auth
{
    public class AuthService : IAuthService
    {
        private readonly IUserRepository _userRepository;
        private readonly JwtUtil _jwtUtil;
        private readonly IConfiguration _configuration;
        private readonly ILogger<AuthService> _logger;

        public AuthService(
            IUserRepository userRepository,
            JwtUtil jwtUtil,
            IConfiguration configuration,
            ILogger<AuthService> logger)
        {
            _userRepository = userRepository;
            _jwtUtil = jwtUtil;
            _configuration = configuration;
            _logger = logger;
        }

        public async Task<UserProfileResponse> GetProfileAsync(string accessToken)
        {
            try
            {
                var decoded = await _jwtUtil.VerifyAsync(accessToken);
                var userId = decoded.Id;
                var user = await _userRepository.FindByIdAsync(userId);
                
                if (user == null || !user.IsEnable)
                {
                    throw new InvalidOperationException("User not found or inactive");
                }
                
                return UserMapper.MapUserToUserProfileResponse(user, _configuration);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error getting user profile");
                throw;
            }
        }

        public async Task<AuthResponse> RefreshTokenAsync(string refreshToken)
        {
            try
            {
                var decoded = await _jwtUtil.VerifyAsync(refreshToken);
                var user = await _userRepository.FindByIdAsync(decoded.Id);
                
                if (user == null || !user.IsEnable)
                {
                    throw new InvalidOperationException("User not found or inactive");
                }

                var tokenPayload = new JwtPayload
                {
                    Id = user.Id,
                    Email = user.Email,
                    Role = user.Role.ToString()
                };
                
                var accessToken = _jwtUtil.GenerateAccessToken(tokenPayload);

                return new AuthResponse
                {
                    AccessToken = accessToken,
                    UserProfile = UserMapper.MapUserToUserProfileResponse(user, _configuration)
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error refreshing token");
                throw;
            }
        }

        public async Task<AuthResponse> AuthenticateAsync(LoginCredentials credentials)
        {
            try
            {
                var user = await _userRepository.FindByEmailAsync(credentials.Email);

                if (user == null)
                {
                    throw new InvalidOperationException("Invalid email or password");
                }

                if (!user.IsEnable)
                {
                    throw new InvalidOperationException("User not found or inactive");
                }

                var isPasswordValid = HashingUtil.VerifyHash(credentials.Password, user.Password, _configuration);

                if (!isPasswordValid)
                {
                    throw new InvalidOperationException("Invalid email or password");
                }

                var tokenPayload = new JwtPayload
                {
                    Id = user.Id,
                    Email = user.Email,
                    Role = user.Role.ToString()
                };

                var accessToken = _jwtUtil.GenerateAccessToken(tokenPayload);
                var refreshToken = _jwtUtil.GenerateRefreshToken(tokenPayload);

                return new AuthResponse
                {
                    UserProfile = UserMapper.MapUserToUserProfileResponse(user, _configuration),
                    AccessToken = accessToken,
                    RefreshToken = refreshToken
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error authenticating user");
                throw;
            }
        }

        public async Task<AuthResponse> CreateAccountAsync(RegisterData registerData)
        {
            try
            {
                _logger.LogInformation("Creating account for email: {Email}", registerData.Email);
                
                var existingUser = await _userRepository.FindByEmailAsync(registerData.Email);
                if (existingUser != null)
                {
                    throw new InvalidOperationException("Email already exists");
                }

                var user = new User
                {
                    Email = registerData.Email,
                    Fullname = registerData.Fullname,
                    Password = registerData.Password,
                    Role = registerData.Role
                };

                var createdUser = await _userRepository.CreateAsync(user);

                if (createdUser != null)
                {
                    var tokenPayload = new JwtPayload
                    {
                        Id = createdUser.Id,
                        Email = createdUser.Email,
                        Role = createdUser.Role.ToString()
                    };

                    var accessToken = _jwtUtil.GenerateAccessToken(tokenPayload);
                    var refreshToken = _jwtUtil.GenerateRefreshToken(tokenPayload);

                    return new AuthResponse
                    {
                        UserProfile = UserMapper.MapUserToUserProfileResponse(createdUser, _configuration),
                        AccessToken = accessToken,
                        RefreshToken = refreshToken
                    };
                }

                throw new InvalidOperationException("An error occurred while creating the account");
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error creating account");
                throw;
            }
        }

        public async Task<User?> GetUserByTokenAsync(string accessToken)
        {
            try
            {
                var decoded = await _jwtUtil.VerifyAsync(accessToken);
                return await _userRepository.FindByIdAsync(decoded.Id);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error getting user by token");
                throw;
            }
        }

        public async Task<List<UserProfileResponse>> GetAllUsersAsync()
        {
            try
            {
                var users = await _userRepository.FindAllAsync();
                var userProfiles = new List<UserProfileResponse>();

                foreach (var user in users)
                {
                    userProfiles.Add(UserMapper.MapUserToUserProfileResponse(user, _configuration));
                }

                return userProfiles;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error getting all users");
                throw;
            }
        }

        public async Task<UserProfileResponse?> UpdateUserAsync(string userId, User updateData)
        {
            try
            {
                var updatedUser = await _userRepository.UpdateAsync(userId, updateData);
                if (updatedUser == null)
                {
                    return null;
                }
                return UserMapper.MapUserToUserProfileResponse(updatedUser, _configuration);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error updating user: {UserId}", userId);
                throw;
            }
        }

        public async Task<UserProfileResponse?> GetUserByEmailAsync(string email)
        {
            try
            {
                var user = await _userRepository.FindByEmailAsync(email);
                if (user == null)
                {
                    return null;
                }
                return UserMapper.MapUserToUserProfileResponse(user, _configuration);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error getting user by email: {Email}", email);
                throw;
            }
        }

        public async Task<UserProfileResponse?> GetUserByIdAsync(string userId)
        {
            try
            {
                var user = await _userRepository.FindByIdAsync(userId);
                if (user == null)
                {
                    return null;
                }
                return UserMapper.MapUserToUserProfileResponse(user, _configuration);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error getting user by ID: {UserId}", userId);
                throw;
            }
        }

        public async Task<UserProfileResponse?> UpdateUserByEmailAsync(string email, User updateData)
        {
            try
            {
                var user = await _userRepository.FindByEmailAsync(email);
                if (user == null)
                {
                    return null;
                }
                
                var updatedUser = await _userRepository.UpdateAsync(user.Id, updateData);
                if (updatedUser == null)
                {
                    return null;
                }
                
                return UserMapper.MapUserToUserProfileResponse(updatedUser, _configuration);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error updating user by email: {Email}", email);
                throw;
            }
        }

        public async Task<bool> DeleteUserByEmailAsync(string email)
        {
            try
            {
                var user = await _userRepository.FindByEmailAsync(email);
                if (user == null)
                {
                    return false;
                }
                return await _userRepository.DeleteAsync(user.Id);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error deleting user by email: {Email}", email);
                throw;
            }
        }

        public async Task<UserProfileResponse?> UpdateUserStatusByEmailAsync(string email, bool isEnable)
        {
            try
            {
                var user = await _userRepository.FindByEmailAsync(email);
                if (user == null)
                {
                    return null;
                }
                
                var updatedUser = await _userRepository.UpdateStatusAsync(user.Id, isEnable);
                if (updatedUser == null)
                {
                    return null;
                }
                
                return UserMapper.MapUserToUserProfileResponse(updatedUser, _configuration);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error updating user status by email: {Email}", email);
                throw;
            }
        }
    }
}

